Enhance the evaluation rules of AccessPolicy objects (#220)
kubernetes-sigs/kube-agentic-networking: docs/proposals/0059-AccessPolicyTargetRefs.md

### Coexistence and Evaluation Flow
It is allowed to have `AccessPolicy` objects targeting a `Gateway` object and `AccessPolicy` objects targeting a `Backend` object behind the `Gateway` object. The evaluation follows a strict hierarchy:

#### 1. Gateway-Level Evaluation (First)
The `AccessPolicy` objects targeting the `Gateway` object are evaluated first.

- **[ExternalAuth-type](https://github.com/kubernetes-sigs/kube-agentic-networking/blob/cf8c85b85d067657a5dce7b87270f8099f1e302c/api/v0alpha0/accesspolicy_types.go#L169) Rules**:
    - Evaluated first.
    - If **any** ExternalAuth rule matching the request returns `deny`, the request is denied immediately.
- **[InlineTools-type](https://github.com/kubernetes-sigs/kube-agentic-networking/blob/cf8c85b85d067657a5dce7b87270f8099f1e302c/api/v0alpha0/accesspolicy_types.go#L165) Rules**:
    - Evaluated only if all ExternalAuth rules (if any) allowed the request.
    - If there are **no** InlineTools rules matching the request, the request is allowed.
    - If **any** InlineTools rule matching the request exists and allows it, the request is allowed.

>     - Otherwise, the request is denied.

Reviewer: Although I'd still prefer the rules to be evaluated in the order they appear in the policy, I think this is an acceptable compromise if it's to keep the [deny-overrides behaviour](https://github.com/kubernetes-sigs/kube-agentic-networking/pull/220#issuecomment-4197844376).

For the reference implementation:

I think this will require configuring two RBAC filters - before and after ExternalAuth. So the config goes:
1. Gateway-level RBAC filter with shadow rules for ExternalAuth
2. Gateway-level Ext_authz filter
3. Gateway-level RBAC filter with normal rules for InlineTools
4. Backend-level RBAC filter with shadow rules for ExternalAuth
5. Backend-level Ext_authz filter
6. Backend-level RBAC filter with normal rules for InlineTools

- **[ExternalAuth-type](https://github.com/kubernetes-sigs/kube-agentic-networking/blob/cf8c85b85d067657a5dce7b87270f8099f1e302c/api/v0alpha0/accesspolicy_types.go#L169) Rules**:
    - Evaluated first.
    - If **any** ExternalAuth rule matching the request returns `deny`, the request is denied immediately.
- **[InlineTools-type](https://github.com/kubernetes-sigs/kube-agentic-networking/blob/cf8c85b85d067657a5dce7b87270f8099f1e302c/api/v0alpha0/accesspolicy_types.go#L165) Rules**:
    - Evaluated only if all ExternalAuth rules (if any) allowed the request.
    - If there are **no** InlineTools rules matching the request, the request is allowed.
    - If **any** InlineTools rule matching the request exists and allows it, the request is allowed.
    - Otherwise, the request is denied.

## Summary Table

The following table summarizes the evaluation results for different combinations of policy checks. Note that evaluation short-circuits as soon as a definitive Deny or Allow is reached.

| Gateway ExternalAuth | Gateway InlineTools | Backend ExternalAuth | Backend InlineTools | Final Result |
| :--- | :--- | :--- | :--- | :--- |
| **Deny** (Any denies) | - | - | - | **Deny** |

> | **Allow** (All allow) | **Deny** (All deny) | - | - | **Deny** |

Reviewer: How can you deny with `InlineTools` today?

1. A single `AccessPolicy` object targeting a Gateway and a Backend at the same time is NOT allowed.

> 1. It is allowed to have `AccessPolicy` objects targeting a `Gateway` object and `AccessPolicy` objects targeting a `Backend` object behind the `Gateway` object. In this case, the `AccessPolicy` objects targeting the `Gateway` object will be evaluated first. Among the `AccessPolicy` objects targeting the `Gateway` object, the [ExternalAuth-type](https://github.com/kubernetes-sigs/kube-agentic-networking/blob/cf8c85b85d067657a5dce7b87270f8099f1e302c/api/v0alpha0/accesspolicy_types.go#L169) authorization rules matching the request, if exist, will be evaluated first, the request will be denied if the evaluation result of any of the ExternalAuth-type authorization rules is `deny`. If there are no [InlineTools-type](https://github.com/kubernetes-sigs/kube-agentic-networking/blob/cf8c85b85d067657a5dce7b87270f8099f1e302c/api/v0alpha0/accesspolicy_types.go#L165) authorization rules matching the request, allow the request. If any InlineTools-type authorization rule matching the request exists, allow the request. Otherwise, deny the request.

Reviewer 2: Because "no rules matching" and "any rules matching" cover 100% of all possible scenarios, the "Otherwise, deny" condition can never be reached. Suggesting a few changes to make this clearer:

```suggestion
1. Separate `AccessPolicy` resources can target a `Gateway` object and a `Backend` object behind that `Gateway` object. In this case, the `AccessPolicy` objects targeting the `Gateway` object will be evaluated first. Among the `AccessPolicy` objects that apply to a request, any ExternalAuth-type authorization rules will be evaluated first. The request will be denied if the evaluation result of any of the ExternalAuth-type authorization rules is `deny`. If any InlineTools-type authorization rule matching the request exists, the request will be allowed. Otherwise, the request will be denied.
```

        BPI2(AccessPolicy: backend-policy-inline-tools-2)
    end

    %% Target arrows
    BPA1 -. "TargetRefs" .-> BE
    BPA2 -. "TargetRefs" .-> BE
    BPI1 -. "TargetRefs" .-> BE
    BPI2 -. "TargetRefs" .-> BE
    
    %% Force BackendPolicyLayer below BE
    BE ~~~ BPA1
    BE ~~~ BPA2
    BE ~~~ BPI1
    BE ~~~ BPI2

    %% --- STYLING ---
    classDef infra fill:#e3f2fd,stroke:#1565c0,stroke-width:2px;
    classDef routing fill:#e8f5e9,stroke:#2e7d32,stroke-width:2px;
    classDef policy fill:#fff3e0,stroke:#ef6c00,stroke-width:2px,stroke-dasharray: 5 5;

    class GW,BE infra;
    class Route routing;
    class GPA1,GPR1,GPA2,GPR2,BPA1,BPA2,BPI1,BPI2 policy;
    style PolicyLayer fill:none,stroke:#ccc,stroke-dasharray: 5 5;
    style BackendPolicyLayer fill:none,stroke:#ccc,stroke-dasharray: 5 5;
```

### Evaluation Flow

When a request comes to `payment-service` through `prod-gateway`:

#### 1. Gateway Level Checks
- **ExternalAuth Policies**: `gateway-policy-external-auth-1` and `gateway-policy-external-auth-2` are evaluated.
    - If **any** denies, the request is rejected immediately.
    - If **all** allow, proceed.

>     - *Note: Evaluation order among these policies does not matter.*

Reviewer 3: Not sure if this "does not matter" is meant to mean "in the way its specified" or "to users"?

I think its probably important to users the order; usually if they are chaining, they are using external auth not just to give a `allow/deny` but to augment the request with additional information.

However I agree its a very challenging problem to solve if we allow multiple across resources 🙂 just that I would expect that eventually we see demand for ordering

Author: > I think its probably important to users the order; usually if they are chaining, they are using external auth not just to give a allow/deny but to augment the request with additional information.

Good point. 

Before this PR, the ordering is: the policies with earlier creationTimestamp will be evaluated first. For the policies with the same creationTimestamp, the ones appearing first in alphabetical order by `{namespace}/{name}` will be evaluated first.  

Maybe we can keep this ordering for both ExternalAuth and ALLOW/InlineTools policies (strictly speaking, the evaluation order of ALLOW policies really does not matter). WDYT?

Reviewer 4: I actually noticed this reading through https://github.com/kubernetes-sigs/kube-agentic-networking/blob/main/docs/proposals/0059-AccessPolicyTargetRefs.md and felt the timestamp behavior was a bit unexpected - in Gateway API timestamps are mostly used to [resolve conflicts](https://gateway-api.sigs.k8s.io/guides/api-design/#conflicts) during the evaluation of a resource by a controller (deciding whether to accept or reject it, such as if only a certain number of Gateway-targeted policies were allowed), and aren't typically relevant during evaluation/enforcement of a resource or policy by the dataplane at runtime.

> First, gateway-policy-audit is evaluated (earlier creation timestamp).
Next, gateway-policy-region is evaluated.

I can see how, as @howardjohn mentions, mutations which augment a request could be relevant for subsequent policy decisions, but I think a more explicit ordered configuration would be preferable there, and I'm not sure if the mutating scope extends beyond current expectations for AccessPolicy and starts looking more like https://github.com/kubernetes-sigs/wg-ai-gateway/blob/main/proposals/7-payload-processing.md, which does have specific expectations around ordering and request mutation?

Author: @mikemorris , thanks for the feedback. Folk seems in favor of limiting the number of ExternalAuth policies/rules per target to 1 (see the GH [discussion](https://github.com/kubernetes-sigs/kube-agentic-networking/discussions/235#discussioncomment-16506417)).

I will update this PR after we finalize the GH dicussion regarding this.

- **InlineTools Policies**: `gateway-policy-inline-tools-1` and `gateway-policy-inline-tools-2` are evaluated.
    - If **any** allows, proceed to Backend level.

>     - If **all** deny, the request is rejected immediately.

Reviewer: Except that the request is _not explicitly denied_ at the gateway level. Rather, it's _not allowed_, which can then lead to _implicitly denied_, or alternatively could lead to _deferred to backend level_.

I wonder if an implicit deny at the gateway level when there are still other policies to check at backend level is the behaviour users expect.

This behaviour is equivalent to the [**atomic overrides**](https://gateway-api.sigs.k8s.io/geps/gep-713/#merge-strategy-atomic-overrides) merge strategy described by GEP-713. AFAIK, it has no counterpart in Istio, it's the exact opposite of the behaviour implemented by Envoy Gateway (atomic defaults), and exists in Kuadrant as an opt-in feature.

>     - If **all** deny, the request is rejected immediately.

Author: > I wonder if an implicit deny at the gateway level when there are still other policies to check at backend level is the behaviour users expect.

I agree that we need user feedback regarding this. 
Before we have user feedback, let us stick to `implicit deny`. WDYT?

@guicassolato  @mikemorris @howardjohn

>     - If **all** deny, the request is rejected immediately.

Reviewer 4: I think the trickiest part here is distinguishing between an explicitly configured empty set `tools: []` field versus an unset/omitted field (which is particularly annoying in Go implementations, and would necessitate using a pointer type).

IMO, [strict hierarchical evaluation](https://github.com/kubernetes-sigs/kube-agentic-networking/issues/55#issuecomment-3818898257) (Gateway before Backend) means:
- `tools: [y]` on a Gateway-targeted policy evaluates to an explicit ALLOW for calling tool `y`, any applicable Backend-targeted policy is evaluated next.
- `tools: [x]` on a Gateway-targeted policy evaluates to an _implicit_ "allow no tools except x" - an attempt to call tool `y` is not allowed, so any Backend-targeted policy is irrelevant
- `tools: []` on a Gateway-targeted policy _implicitly_ allows no tool calls, any Backend-targeted policy is irrelevant

Critically, the "implicit allow nothing" is _not_ actually modeled as a DENY which would halt further evaluation and reject the request, it's more like implicitly adding a higher-level default cluster-wide "allow nothing". In Istio this is actually [created explicitly as an "allow-nothing" policy](https://istio.io/latest/docs/reference/config/security/authorization-policy/), whereas NetworkPolicy enforces this implicitly when an ALLOW rule is created (similar to what's being proposed here).

```
kind: XAccessPolicy
spec:
  targetRefs:
  - kind: Gateway
  rules: []
```
^ (or similarly, the lack of any AccessPolicy targeting the Gateway) _does not specify_ any InlineTools ALLOW rules, and thus I would expect no implicit deny should exist and any Backend-targeted should be evaluated. Backend-targeted policies would also be evaluated when a Gateway-targeted AccessPolicy _does_ explicitly allow a requested tool.

This is a somewhat contrived case currently, but could be much more relevant if AccessPolicy eventually became an additional ruleset extension of [AuthorizationPolicy](https://gateway-api.sigs.k8s.io/geps/gep-3779/) (which would make broader L4/L7 rules omitting a `tools` stanza much more common) or added an audit/"shadow" mode (for logging/testing before enforcement) as below:
```
  rules:
  - authorization:
      type: InlineTools
      action: ALLOW
      mode: AUDIT | ENFORCED (default when mode field is omitted)
      tools: 
      - x
```
(I've always found AUDIT as a [peer enum value alongside ALLOW and DENY](https://istio.io/latest/docs/reference/config/security/authorization-policy/#AuthorizationPolicy-Action) as confusing, because the intent of the rule becomes less clear.)

If we add a DENY action, we should consider whether that would also trigger the implicit "allow nothing" layer or not (as a narrow DENY breaking all traffic might be unexpected).

    - *Note: Evaluation order among these policies does not matter.*

#### 2. Backend Level Checks
- **ExternalAuth Policies**: `backend-policy-external-auth-1` and `backend-policy-external-auth-2` are evaluated.
    - If **any** denies, the request is rejected immediately.
    - If **all** allow, proceed.
    - *Note: Evaluation order among these policies does not matter.*
- **InlineTools Policies**: `backend-policy-inline-tools-1` and `backend-policy-inline-tools-2` are evaluated.
    - If **any** allows, the request is allowed.

>     - If **all** deny, the request is denied.

Reviewer: ```suggestion
    - If **none** allow, the request is denied.
```

| High | `HTTPRoute` / `GRPCRoute` | 2 |
| Medium | `Gateway` Listener (via `sectionName`) | 3 |
| Lowest | `Gateway` (Entire resource) | 4 |

#### Handling Conflicts (Same Level)

If two different `SecurityPolicy` objects target the exact same resource (e.g., two policies both targeting `HTTPRoute: my-service`), Envoy Gateway uses "Tie-breaking" rules:

* Oldest Wins: The policy with the earliest `creationTimestamp` is applied.

* Alphabetical Sorting: If the timestamps are identical, the policy whose name comes first alphabetically wins.

* Status Reporting: The "losing" policies will show a status of Overridden=True or Conflicted=True in their Kubernetes status field.

### Linkerd Authorization Policy

[Linkerd Authorization Policy](https://linkerd.io/2-edge/reference/authorization-policy/) evaluation follows a "Specific-to-General" hierarchy that centers around the concept of a `Server` resource.

Linkerd's proxy makes an authorization decision for every incoming request using these checks in order:

1. Check for a `Server` resource:

    * NO `Server` exists: The proxy uses the Default Policy (see below).

    * YES `Server` exists: Proceed to step 2.

> 1. Check for an `HTTPRoute`:

Reviewer: nit:

```suggestion
2. Check for an `HTTPRoute`:
```
